### ISSUE TITLE
$unset at discriminated model can remove the discriminator key from documents

### ISSUE THREAD

**Issue by borfig** [Automattic/mongoose#11456]
**Do you want to request a *feature* or report a *bug*?**
*bug*

**What is the current behavior?**
Using a discriminated model, it is possible to `$unset` the discriminator key field.

**If the current behavior is a bug, please provide the steps to reproduce.**
This short script demonstrates the issue:
```javascript
"use strict";
// shamelessly copying the example from the docs
const { connect, Schema, model, connection } = require("mongoose");
const options = { discriminatorKey: 'kind' };
const eventSchema = new Schema({ time: Date }, options);
const Event = model('Event', eventSchema);
const ClickedLinkEvent = Event.discriminator('ClickedLink',
  new Schema({ url: String }, options));
async function main() {
  try {
    await connect("mongodb://mongo:27017");
    await Event.deleteMany({});
    await ClickedLinkEvent.create({
      time: new Date(),
      url: "http://www.example.com",
    });
    const e = await Event.findOne();
    console.log(e?.kind); // output: ClickedLink
    try {
      await ClickedLinkEvent.updateMany({
        $set: {
          kind: "SomethingElse"
        }
      });
    } catch (error) {
      console.log(error.message); // output: Can't set discriminator key "kind"
    }
    const eu = await ClickedLinkEvent.findOneAndUpdate({
      url: "https://mongoosejs.com/",
    }, {
      $setOnInsert: {
        kind: "SomethingElse",
        time: new Date(),
      }
    }, {
      new: true,
      upsert: true,
    });
    console.log(eu?.kind); // output: ClickedLink
    await ClickedLinkEvent.updateMany({
      $unset: {
        kind: "",
      }
    });
    const e2 = await Event.findOne();
    console.log(e2?.kind ?? "MISSING"); // output: MISSING
    const e3 = await ClickedLinkEvent.findOne();
    console.log(e3 ? "found" : "not found"); // output: not found
  } finally {
    await connection.close();
  }
}
void main();
```
`$set` throws an error, and `$setOnInsert` does nothing. However `$unset` is not blocked.

**What is the expected behavior?**
When using a discriminated model (`ClickedLinkEvent` in the example above), `$unset` of a discriminator key should be either blocked (with error as `$set` does) or ignored (as `$setOnInsert` does).

However, it is OK to allow `$unset` of discriminator key at the global model (`Event`).

**What are the versions of Node.js, Mongoose, and MongoDB you are using? Note that "latest" is not a version.**
- Node.js v16.14.0
- Mongoose  v6.2.3
- MongoDB v5.0.6

**Comment by IslandRhythms** [comment]
```javascript

"use strict";
// shamelessly copying the example from the docs
const mongoose = require("mongoose");
const {Schema, connect, connection, model} = mongoose;
const options = { discriminatorKey: 'kind' };
const eventSchema = new Schema({ time: Date }, options);
const Event = model('Event', eventSchema);
const ClickedLinkEvent = Event.discriminator('ClickedLink',
  new Schema({ url: String }, options));
async function main() {
  try {
      console.log('start');
    await connect("mongodb://localhost:27017");
    await connection.dropDatabase();
    await ClickedLinkEvent.create({
      time: new Date(),
      url: "http://www.example.com",
    });
    const e = await Event.findOne();
    console.log(e.kind); // output: ClickedLink
    try {
      await ClickedLinkEvent.updateMany({
        $set: {
          kind: "SomethingElse"
        }
      });
    } catch (error) {
      console.log(error.message); // output: Can't set discriminator key "kind"
    }
    const eu = await ClickedLinkEvent.findOneAndUpdate({
      url: "https://mongoosejs.com/",
    }, {
      $setOnInsert: {
        kind: "SomethingElse",
        time: new Date(),
      }
    }, {
      new: true,
      upsert: true,
    });
    console.log(eu?.kind); // output: ClickedLink
    await ClickedLinkEvent.updateMany({
      $unset: {
        kind: "",
      }
    });
    const e2 = await Event.findOne();
    console.log(e2.kind ? e2.kind : "MISSING"); // output: MISSING
    const e3 = await ClickedLinkEvent.findOne();
    console.log(e3 ? "found" : "not found"); // output: not found
  } finally {
    await connection.close();
    console.log('done');
  }
}
void main();
```

**Comment by AbdelrahmanHafez** [comment]
I think it's a valid use case to allow updating the `kind` manually to convert one document from one discriminator to another.
Of course that comes with the risks of keeping documents in an invalid state, but that's the responsibility of the developer.

I'm leaning more towards throwing an error _only_ if the `kind` is not a valid discriminator name, otherwise, I think we should allow people to update the kind if they want to.

**Comment by borfig** [comment]
IMHO changing `kind` should be OK only on the non-discriminated model. The discriminator should treat it as read-only.

In the above example, I do believe that:
```javascript
await Event.updateOne({/*snip*/}, { $set: { kind: "SomeOtherKind"} });
```
is a valid operation. However,
```javascript
await ClickedLinkEvent.updateOne({/*snip*/}, { $set: { kind: "SomeOtherKind"} });
```
should throw an error, and -
```javascript
await ClickedLinkEvent.updateOne({/*snip*/}, { $set: { kind: "ClickedLink"} });
```
could do nothing (today it is throwing an error as well, no?)

For now, let's focus on `$unset` - the subject of this issue.

**Comment by AbdelrahmanHafez** [comment]
As far as `$unset`, I agree that we should block this operation for discriminator models.

**Comment by vkarpov15** [comment]
Fix will be in 6.2.5. You won't be able to update the discriminator key using update operators like `$unset` or `$push` _unless_ you set the `overwriteDiscriminatorKey` option.
